spec/compiler/semantic/automatic_cast.cr

```diff
@@ -0,0 +1,370 @@
+require "../../spec_helper"
+
+describe "Semantic: automatic cast" do
+  it "casts literal integer (Int32 -> no restriction)" do
+    assert_type(%(
+      def foo(x)
+        x + 1
+      end
+
+      foo(12345)
+      ), inject_primitives: true) { int32 }
+  end
+
+  it "casts literal integer (Int32 -> Int64)" do
+    assert_type(%(
+      def foo(x : Int64)
+        x
+      end
+
+      foo(12345)
+      )) { int64 }
+  end
+
+  it "casts literal integer (Int64 -> Int32, ok)" do
+    assert_type(%(
+      def foo(x : Int32)
+        x
+      end
+
+      foo(2147483647_i64)
```

Comment:
Why does this work?
I would expect that if the literal is explicitely typed, that type should be honored

Reply:
Yeah, maybe that shouldn't work, but I don't know, there's no harm in casting the value.

Plus, right now there's no concept of a number that's **explicitly** typed. Or, well, there's no difference between `1` and `1_i32`, both parse to exactly the same thing, and are handled in the same way in the compiler. So maybe we should start making that distinction and only apply the rule for values that don't have a suffix...

It should be easy to implement, I might try that tomorrow. But I don't think this particular thing should block the entire feature from being merged.

Reply:
Seems okay as long as type on a method or C fun is honored.